### ISSUE TITLE
Screen garbling: xvi emits cursor motions beyond the last screen column

### ISSUE THREAD

**Issue by martinwguy** [martinwguy/xvi#71]
When you insert characters into a line 79 characters long in the middle of the screen, the inserted text is garbled on the screen.

Example: Changing cur to current ("_" marks the cursor position, ">" typed keys):
Line at start of test (78 characters):
```
 * would have had to start playing uninterruptedly to be at the cur_point now.
```
> ir
```
 * would have had to start playing uninterruptedly to be at the curr_point now.
```
This is OK.
> e
```
 * would have had to start playing uninterruptedly to be at the curre point now.
```
! The cursor is on the e !
> n
```
 * would have had to start playing uninterruptedly to be at the currn point now.
.
```
!! The n's in the wrong place and the cursor is on the p !!
> t
```
 * would have had to start playing uninterruptedly to be at the currn t point no
w.
```

When it overflows the line at 'n', albeit with garbage, it correctly scrolls the rest
of the screen down one line. Esc ^R repaints the screen as it should be.

**Comment by jvvv** [comment]
In trying to reproduce this in different ways, I am finding that the number of columns in the line does not seem to be to key. Rather, the point at which the bug seems to crop up seems to be related to columns from the end to the row. At least that is my impression when trying to reproduce. Another thing to note is that this bug does not seem to manifest when linked against termcap rather than curses... at least so far.

**Comment by jvvv** [comment]
Ok, bug occurs with both termcap and curses libraries. But the bug manifests at different columns depending on which library is linked to.

**Comment by martinwguy** [comment]
If you can give some specific tests that do/do not exhibit the bug,
that might help narrow down the cause of the bug. Borderline cases,
where a tiny change makes it come and go, are particularly
interesting.

Of course, it may depend on the range of terminal capabilities
available - the above tests were with xterm.

**Comment by martinwguy** [comment]
Here's a small one:
```
8a1234567890[Esc]xiX[Esc]
```
should give
```
123456789012345678901234567890123456789012345678901234567890123456789012345678X9
```
and the cursor should be on the X.
So far, xvi gets the display right but the cursor is on the 8.

Capturing the output with ```script -f``` show that the character sequence it uses to
position the cursor is ```^[[1;81H^H^H```: move to column 81 and go left two spaces!

In fact, reviewing the typescript file, it often outputs "go to column 81", elsewhere
followed immediately by ^M.

**Comment by jvvv** [comment]
I have noticed that anytime the end of line is at the right most column is when cursor positioning and display seems to start going awry. I have a simple patch, [xvi_issue71_workaround.txt](https://github.com/martinwguy/xvi/files/706719/xvi_issue71_workaround.txt), that seems clears up the issue temporarily. I don't think that this is a long term solution, just confirmation of where the trouble manifests. I'm not convinced that this is the only trouble spot, as there is reason to believe that the two routines in cursor.c may contribute to the issue because these 'fixup' the cursor position. Looks like you were about on top of the issue and I think the work around above confirms it.

**Comment by jvvv** [comment]
There are two things going wrong that cause this issue. First, there is a missed conditional else in xyupdate. Second, the cursor does not consistently get updated. The changes are minimal. I saw some relate stuff I'd like to clean up, but will save that for later. Will have the branch pushed and a PR for it shortly.

**Comment by martinwguy** [comment]
Using "git bisect" on xterm, the defect starts happening at commit 896dab99
```
    Remove s_inschar for i_proc.
    
    The s_inschar call was redundant, since replchars call earlier
    in i_proc handles adding the char to the screen buffer.
--- a/src/edit.c
+++ b/src/edit.c
@@ -385,7 +385,6 @@ int c;
      * Update the screen.
      */
     if (!updated) {
-       s_inschar(curwin, c);
        updateline(curwin);
    
```
which is odd because the comment for s_inschar() says
```
/*
 * Note that this routine doesn't have to do anything, as
 * the screen will eventually be correctly updated anyway; it's just
 * here for speed of screen updating.
 */
```
And, in fact, in a lot of cases s_inschar() just returns without outputting or changing anything. It could be that, if s_inschar() does the on-the-fly screen update, then updateline() doesn't have as much to do when it is called and so a buggy code path in updatelines() is not called.

Incidentally, the comment for the commit is not quite right. replchars() adds the character to the edit buffer but doesn't update the screen. s_inschar() might update the screen or might not, but it seems to be the only opportunity xvi has to use the "insert character" or "insert mode" capabilities to shunt the rest of the line to the right, while, if s_inschar() isn't called, updatelines() finds that the two lines are all different and so repaints the whole line for every insertion.
Let's test that theory:
Test piece: "8aa^[Ibbbbbh"
With s_inschar():
```
[...]^M^[[4hb^[[4l^[[4hb^[[4l^[[4hb^[[4l[...]
```
without:
```
[...]^Mbaaaaaaaa^H^H^H^H^H^H^H^H^Mbbaaaaaaaa^H^H^H^H^H^H^H^H^Mbbbaaaaaaaa
```
Of course, neither of these update sequences is optimal (staying in insert mode might be tricky to implement but using all those ^H's instead of CM or ^M ND or ^Mb is just bizarre!), but I think that explains the usefulness of a call to s_inschar() here and I've reinstated it in commit 83192720e

I've made a test in the testsuite, 0071-*, but it never fail, probably because tkterm and virterm's termcaps don't have an insert mode or insert character capability, so a different code path is followed.

There's still a bug in updateline() here, which should update the line correctly whether s_inschar() is called or not, but reinstating s_inschar() here papers over it.
The bug in updateline() remains, but to see it now using the above tests you first need to remove the call to s_inschar() in edit.c

**Comment by martinwguy** [comment]
Fixed by 83192720

**Comment by martinwguy** [comment]
This still happens. It's hard to reproduce but happens when inserting near the end of a line makes the line wrap.

**Comment by martinwguy** [comment]
May be fixed by 0a872b9e, a cursor motion/colour optimization.
At least, the go-to-column-81 behaviour is eliminated by that.
